### PR TITLE
feat(sbom): optimize C# call parsing and bindings

### LINKED ISSUE #213: [C# Support] Wire C# parsing into the SBOM extraction pipeline
## Expected Behavior

`nuguard sbom generate` recognizes `.cs` source files, runs the C# parser and framework adapters against them, and merges the resulting nodes/edges into the `AiSbomDocument` alongside Python/JS/Go output.

## Current Behavior

`.cs` is not in `AiSbomConfig.include_extensions` (`nuguard/sbom/config.py`), so `extractor/core.py`'s source-file walk (`core.py:2450`) skips C# files entirely — this is the config change that does *not* belong in #210 (dependency scanning walks manifests unconditionally by filename, independent of this setting).

## Possible Solution

1. Add `.cs` (and `.csx` if in scope) to `AiSbomConfig.include_extensions`.
2. Wire the `csharp_parser.py` output and adapters from #211/#212 into `extractor/core.py`'s per-file processing, following the existing pattern for how Python/JS parse results feed into node/edge construction.
3. Confirm merged output round-trips through the existing SBOM schema/merger without special-casing (`nuguard/sbom/schemas/aibom.schema.json`, `test_committed_schema_matches_models`).

## Context

Part of #170. Depends on #210, #211, and #212 all landing first — this is the integration point that makes the C# work actually reachable from the CLI.

### LINKED ISSUE #260: [C# Support] Harden C# framework adapter edge cases before pipeline wiring
## Expected Behavior

The C# framework adapters should preserve C# invocation, routing, request-shape, provider, model, and plugin semantics across common .NET usage patterns:

* Fluent invocation chains such as `Trainer().Fit(...)` and `CreateBuilder().Build()` should preserve the preceding invocation as the receiver and retain the final assignment target.
* ASP.NET Core controller actions should support:

  * a separate method-level `[Route(...)]` combined with `[HttpPost]`, `[HttpGet]`, and related HTTP attributes
  * absolute action templates beginning with `/` or `~/`, which override the controller route
  * collection-valued conversational fields such as `List<Message>` and `string[]`
  * explicitly body-bound primitive prompts such as `[FromBody] string prompt`
* An unqualified legacy `OpenAIClient` should be classified as Azure OpenAI when `Azure.AI.OpenAI` is imported and the OpenAI namespace is not, even when unrelated providers are also imported.
* Semantic Kernel should:

  * avoid creating a fabricated model node when a model or deployment expression cannot be resolved
  * resolve positional plugin names according to the registration method and overload being used

## Current Behavior

A follow-up review of #222 identified these edge cases:

* `find_calls()` only records the identifier chain immediately before the current `(`. In `Trainer().Fit(...)`, `Fit` therefore loses its receiver and the final assignment target.
* A method-level `[Route("complete")]` is ignored when the HTTP attribute has no inline template.
* Absolute ASP.NET action routes are concatenated with the controller route instead of replacing it.
* `chat_payload_list` is always emitted as `false`, including for array and collection fields.
* Primitive parameters are discarded even when marked `[FromBody]` or named like conversational input.
* Legacy `OpenAIClient` is treated as Azure only when Azure is the sole imported provider.
* An unresolved Semantic Kernel model registration derives a model name from the registration method, producing values such as `AzureOpenAI` that are not model identifiers.
* Semantic Kernel plugin extraction always assumes positional argument zero is the plugin name, although object and prompt-directory registrations place it later.

## Possible Solution

* Parse all call expressions before applying the optional name filter.
* Link a call to the immediately preceding invocation when the expressions are joined by `.` or `?.`.
* Inherit the fluent chain’s assignment target and preserve the entire receiver expression.
* Ensure ML.NET `Fit` processing runs after trainer and transform recognition without treating `Fit` itself as the trainer or transform call.
* Fall back to a method-level `[Route]` when the HTTP attribute has no inline route.
* Treat action routes beginning with `/` or `~/` as absolute.
* Derive `chat_payload_list` from the request-field type.
* Preserve body-bound and conversationally named primitive parameters and synthesize their request schema.
* Classify legacy `OpenAIClient` as Azure when Azure is imported and OpenAI is not.
* Skip unresolved Semantic Kernel model registrations instead of fabricating a model identifier.
* Use registration-specific positional candidates for Semantic Kernel plugin names.
* Add focused regression tests for every behavior above.

## Context

This is a follow-up to #212 and PR #222.

The adapters are currently additive-only and are not yet registered in the normal extraction pipeline. This hardening work should land before #213 wires the adapters into regular AI-SBOM generation.

Out of scope:

* `.cs` and `.csx` source configuration
* adapter registry changes
* extraction-pipeline wiring
* C# Semgrep rules
* end-to-end C# fixture applications
* runtime behavior or red-team changes

### PR DESCRIPTION
**Pull request by fatkobra**
Closes #260 

Follow-up to #260 and #290.
Supports the C# pipeline integration tracked in #213.
Part of #170.

## PR Type

* [ ] Bug fix
* [x] Feature

## What

This PR addresses the two non-blocking follow-up findings identified during the maintainer review of #290.

It:

* replaces reverse linear fluent-chain parent lookup with an offset-indexed lookup
* avoids parsing call arguments for calls excluded by a supplied `names` filter
* preserves fluent receiver and assignment-target behavior
* preserves explicit ASP.NET Core parameter-binding sources
* prevents query-, route-, form-, header-, service-, and aggregate-bound parameters from being represented as JSON request-body fields
* prevents unattributed primitive Minimal API parameters from being represented as request-body fields
* preserves explicit `[FromBody]` primitive request metadata
* preserves existing complex request-DTO behavior
* adds focused scalability and binding-source regression tests

## Why

The C# call parser introduced fluent-chain reconstruction by retaining every call and searching backward through all previously parsed calls for each candidate.

That correctly preserves expressions such as:

```csharp
var model = trainer.Build().Fit(data);
```

but the reverse scan makes chain-parent lookup worst-case O(n²) over call count. It also causes calls outside a narrow adapter-supplied `names` filter to undergo full argument parsing.

The ASP.NET Core request-selection logic also reduced parameter binding to a `from_body` Boolean and accepted primitive parameters based on conversational names such as `prompt`, `message`, or `query`.

This could incorrectly represent the following as JSON request bodies:

```csharp
[FromQuery] string prompt
```

```csharp
[FromRoute] string message
```

```csharp
[FromForm] string input
```

```csharp
app.MapPost(
    "/chat",
    (string prompt) => ...);
```

These findings were non-blocking for #290 because the adapters are not yet wired into normal extraction, but they should be addressed before #213 activates the adapters in the standard AI-SBOM pipeline.

### Indexed fluent-chain lookup

The call parser now maintains an index keyed by each parsed call’s end offset.

For each subsequent call, it:

1. scans backward over only adjacent whitespace
2. recognizes a local `.` or `?.` chain separator
3. finds the preceding expression-end offset
4. retrieves the parent call directly from the index

This removes the reverse scan over all prior calls while retaining:

* chained receivers
* null-conditional chains
* assignment-target inheritance
* source snippets
* line information

### Filter-aware argument parsing

All calls continue to be indexed structurally because an unselected call may be the parent of a selected call.

However, `parse_arguments()` is now invoked only when:

* no name filter was supplied, or
* the call’s simple method name is included in the supplied filter

This keeps fluent reconstruction intact without fully parsing unrelated call arguments.

### ASP.NET binding-source preservation

Parameter parsing now distinguishes these explicit binding sources:

* `body`
* `query`
* `route`
* `form`
* `header`
* `services`
* aggregate `parameters`

The following attributes are recognized:

* `[FromBody]`
* `[FromQuery]`
* `[FromRoute]`
* `[FromForm]`
* `[FromHeader]`
* `[FromServices]`
* `[FromKeyedServices]`
* `[AsParameters]`

Fully qualified attribute names and `Attribute` suffixes are normalized before matching.

### Request-body selection

Request-body metadata is now limited to:

* explicitly body-bound primitive parameters
* explicitly body-bound complex DTOs
* complex DTOs without an explicit non-body binding source

These are excluded from `request_body_schema`:

* query-bound parameters
* route-bound parameters
* form-bound parameters
* header-bound parameters
* service-injected parameters
* aggregate `[AsParameters]` parameters
* unattributed primitive parameters

A conversational name alone no longer changes a parameter’s binding semantics.

## Test Coverage

Regression tests cover:

* preserving a selected fluent-chain receiver
* preserving a fluent-chain assignment target
* skipping argument parsing for 250 unrelated calls when a narrow method-name filter is supplied
* `[FromQuery] string prompt`
* `[FromRoute] string prompt`
* `[FromForm] string prompt`
* `[FromHeader] string prompt`
* `[FromQuery]` complex DTOs
* unattributed primitive Minimal API parameters
* explicit `[FromBody]` primitive Minimal API parameters
* continued compatibility with the existing `[FromBody]` controller test

## Test Steps

* [x] `uv run ruff format --check nuguard/sbom/adapters/csharp/_source.py nuguard/sbom/adapters/csharp/aspnet_core.py nuguard/sbom/tests/test_csharp_framework_adapters.py`
* [x] `uv run ruff check nuguard/`
* [x] `uv run mypy nuguard/`
* [x] `uv run pytest nuguard/sbom/tests/test_csharp_framework_adapters.py -v`
* [x] `uv run pytest nuguard/ -q`
* [x] `uv run pytest tests/ --ignore=tests/apps/contact-center-ai-samples -v`
* [x] `make precommit-run`
* [x] `git diff --check`

## Checks

* [x] The feature branch was created from the latest `develop`
* [x] This PR targets `develop`
* [x] Changed Python files are Ruff-formatted
* [x] Ruff lint passes
* [x] Mypy passes
* [x] Focused C# adapter tests pass
* [x] Package tests pass
* [x] Integration tests pass
* [x] Added regression coverage for both review findings
* [x] No new runtime dependency was added
* [x] No adapter registry changes were added
* [x] No extraction-pipeline wiring was added
* [x] No `.cs` or `.csx` source configuration was added

## Other Notes

This PR intentionally does not implement or close #213.

The following remain out of scope:

* adding `.cs` or `.csx` to default source extensions
* registering C# adapters in the default registry
* extraction-pipeline dispatch
* C# namespace-to-NuGet dependency association
* C# Semgrep security rules
* fixture applications
* runtime behavior or red-team changes

The pipeline integration remains tracked in #213.

**Release note:** Improved C# adapter scalability and ASP.NET Core request-binding metadata accuracy.

### CODE REVIEW

File: nuguard/sbom/adapters/csharp/aspnet_core.py [hunk 1]
```diff
@@ -82,6 +82,16 @@
     "IHeaderDictionary",
     "IServiceProvider",
 }
+_BINDING_SOURCES = {
+    "AsParameters": "parameters",
+    "FromBody": "body",
+    "FromForm": "form",
+    "FromHeader": "header",
+    "FromKeyedServices": "services",
```
[the remark this answers is not in the document]
Comment (KanishkThamman) [reply nuguard/sbom/adapters/csharp/aspnet_core.py:92]: @fatkobra could you address this issue?

[the remark this answers is not in the document]
Comment (fatkobra) [reply nuguard/sbom/adapters/csharp/aspnet_core.py:92]: Fixed in the latest commit.

The root cause was `_lambda_parameters()` using a flat `[^()]*` regex for the complete lambda parameter list. Attribute arguments such as `[FromKeyedServices("chat")]` contain nested parentheses, so the match failed and all parameters—including a valid body DTO beside the service parameter—were discarded.

The implementation now:

* locates the outer lambda parameter list using balanced parenthesis matching
* masks strings and comments while preserving source offsets
* passes the recovered parameter text through the existing nesting-aware `parse_arguments()` helper
* preserves service-bound parameters while still selecting a real `[FromBody]` DTO

A regression test now covers a Minimal API lambda containing both `[FromKeyedServices("chat")] ChatClient client` and `[FromBody] ChatRequest request`.

**Comment by fatkobra** [comment]
Thanks, @KanishkThamman. I addressed the review finding and pushed the update to #320.

The root cause was the Minimal API lambda parser using a flat parameter regex that stopped at the first nested parenthesis. This caused attributes with arguments, such as `[FromKeyedServices("chat")]`, to suppress the entire lambda parameter list and drop any legitimate request DTO beside them.

The update now:

* parses the outer lambda parameter list with balanced delimiters
* supports attribute arguments, default expressions, tuple types, and other nested parameter syntax
* preserves a body DTO when keyed-service parameters are also present
* normalizes C# attribute target specifiers such as `[param: FromQuery]`
* prevents target-qualified query parameters from being represented as request bodies

Regression coverage was added for:

* `[FromKeyedServices("chat")]` together with a real `[FromBody]` DTO
* `[param: FromQuery]` on a complex controller parameter

Validation completed successfully:

* focused C# framework-adapter tests
* Ruff formatting and linting
* mypy
* full package test suite
* integration test suite
* pre-commit hooks
* Git whitespace checks

The review conversation is ready for another look.

**Comment by fatkobra** [comment]
@KanishkThamman @ganesh-nuguard 

The failed package-test job appears to be an unrelated timing-sensitive CI failure.

The failing test was:

`nuguard/behavior/tests/test_v3_efficiency.py::test_build_scenarios_parallelises_llm_calls`

It exceeded the 150 ms wall-clock threshold on the shared runner, while the other 2,279 package tests passed. PR #320 does not modify the behavior scenario implementation or this test.

I reran the exact failing test locally five consecutive times:

```text
Run 1: passed
Run 2: passed
Run 3: passed
Run 4: passed
Run 5: passed
```
I attempted to rerun the failed Actions job through gh, but contributors do not have the required repository permission.

Could you please rerun the failed job?

**Review by ganesh-nuguard — APPROVED** [review]
Looks good
